File: src/com/home/Chapter_4_Stacks_and_Queues/Queue/PriorityQ.kt
package com.home.Chapter_4_Stacks_and_Queues.Queue

class PriorityQ (val maxSize: Int) {
    val queArray = LongArray(maxSize)
    var nItems = 0

    fun insert(value: Long) {
        if (nItems == 0) {
            queArray[nItems++] = value
        } else {
            var i: Int = nItems - 1
            while (i >= 0 && value > queArray[i]) {
                queArray[i + 1] = queArray[i]
                --i
            }
            queArray[i + 1] = value
            ++nItems
        }
    }

    fun remove() = queArray[--nItems]

    fun peekMin() = queArray[nItems - 1]

    fun isEmpty() = nItems == 0

    fun isFull() = nItems == maxSize
}

fun main(args: Array<String>) {
    val queue = PriorityQ(5).apply {
        insert(30L)
        insert(50L)
        insert(10L)
        insert(40L)
        insert(20L)
    }

    while (!queue.isEmpty()) {
        val var2: Long = queue.remove()
        print("$var2 ")
    }

    println("")
}

File: src/com/home/Chapter_14_Weighted_Graphs/scala/PriorityQ.scala
package com.home.Chapter_14_Weighted_Graphs.scala

import scala.util.control.Breaks._

class PriorityQ {
  private val SIZE = 20
  val queArray = new Array[Edge](SIZE)
  var size = 0

  def insert(item: Edge): Unit = {
    var k = 0
    breakable {
      for (i <- 0 until size) {
        k = i
        if (item.distance >= queArray(i).distance) {
          break
        }
      }
    }

    for (i <- size - 1 to k by -1) {
      queArray(i + 1) = queArray(i)
    }

    queArray(k) = item
    size += 1
  }

  def removeMin(): Edge = {
    size -= 1
    queArray(size)
  }

  def removeN(n: Int): Unit = {
    for (i <- n until size - 1) {
      queArray(i) = queArray(i + 1)
    }
    size -= 1
  }

  def peekMin(): Edge = queArray(size - 1)

  def isEmpty: Boolean = size == 0

  def peekN(n: Int): Edge = queArray(n)

  def find(findDex: Int): Int = {
    for (i <- 0 until size) {
      if (queArray(i).destVert == findDex) {
        return i
      }
    }
    -1
  }

}
